tempodb/encoding/vparquet/block_traceql.go

```diff
@@ -0,0 +1,1018 @@
+package vparquet
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"reflect"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/segmentio/parquet-go"
+
+	tempo_io "github.com/grafana/tempo/pkg/io"
+	"github.com/grafana/tempo/pkg/parquetquery"
+	"github.com/grafana/tempo/pkg/traceql"
+)
+
+// Helper function to create an iterator, that abstracts away
+// context like file and rowgroups.
+type makeIterFunc func(columnName string, predicate parquetquery.Predicate, selectAs string) parquetquery.Iterator
+
+const (
+	columnPathTraceID                  = "TraceID"
```

Comment:
we are doing similar things here and this PR: 

https://github.com/grafana/tempo/pull/1678/files#diff-aae971dafc9cc2527a4b439bb75059ea397e6c72de2292b26d875c456f1742c0R39

We should merge one and consolidate this if possible.

Reply:
Yep, the structure here contains some additional information needed for traceql including scope and data type.  I propose that we merge #1678 first and then consolidate on this larger structure since it will work for both purposes. The only concern is if there is a more generic way to represent the data type.  Here is uses traceql types which makes sense, but less so for tag-based search.

Reply:
I think for now we should leave the structures separate and eventually retire the tag-based one.  The main sticking point is the ability of traceql to distinguish between attributes and intrinsics with the same label.  The query `{ .name = 'foo' }` looks for `name` in the generic key/value columns, but tag-based search directs `name=foo` to the span name column.  Couldn't think of a non-hacky way to make one lookup table work for both.  `{ .status.code=x }` has the same problem.   A minor concern was also making tag-based search use traceql constants.